algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala

```diff
@@ -0,0 +1,23 @@
+package com.twitter.algebird
+
+import org.scalatest._
+
+import org.scalatest.prop.PropertyChecks
+import org.scalacheck.{ Gen, Arbitrary, Prop, Properties }
+
+class NumericAlgebraTests extends CheckProperties {
+  import BaseProperties._
+
+  def testRingAndBelowProperties[A: Arbitrary: Ring](name: String): Unit =
+    property(s"test $name") {
+      // check resolution:
+      implicitly[Semigroup[BigDecimal]]
+      implicitly[Monoid[BigDecimal]]
+      implicitly[Group[BigDecimal]]
+      // now test all the laws:
+      ringLaws[A]
+    }
+
+  testRingAndBelowProperties[Int]("int")
+  testRingAndBelowProperties[BigDecimal]("bigdecimal")
```

Comment:
This one keeps failing for me with errors like ```
[info] NumericAlgebraTests:
[info] - test int
[info] - test bigdecimal *** FAILED ***
[info]   GeneratorDrivenPropertyCheckFailedException was thrown during property evaluation.
[info]    (CheckProperties.scala:12)
[info]     Falsified after 0 successful property evaluations.
[info]     Location: (CheckProperties.scala:12)
[info]     Occurred when passed generated values (
[info]       arg0 = 1.627209798210513,
[info]       arg1 = 606303697535274687,
[info]       arg2 = -0.003402952952309675094511817686947572
[info]     )
[info]     Label of failing property:
[info]       isDistributiveDifferentTypes
```

Reply:
it may not be distributive

did you check?

Float and Double are not.

Reply:
when I test with
```
scala> def foo[T](a: T, b: T, c: T)(implicit rng: Numeric[T]) = {
     |     Equiv[T].equiv(rng.times(a, rng.plus(b, c)), rng.plus(rng.times(a, b), rng.times(a, c))) &&
     |     Equiv[T].equiv(rng.times(rng.plus(b, c), a), rng.plus(rng.times(b, a), rng.times(c, a)))
     | }
```
it seems to work with the failing values as BigDecimal

Reply:
ahhh... it's rounding errors
```
scala> val a = BigDecimal(1.627209798210513)
a: scala.math.BigDecimal = 1.627209798210513

scala> val b = BigDecimal("606303697535274687")
b: scala.math.BigDecimal = 606303697535274687

scala> val c = BigDecimal("-0.003402952952309675094511817686947572")
c: scala.math.BigDecimal = -0.003402952952309675094511817686947572
scala> def bar[T](a: T, b: T, c: T)(implicit rng: Numeric[T]) = {
     | (rng.times(a, rng.plus(b, c)), rng.plus(rng.times(a, b), rng.times(a, c)), rng.times(rng.plus(b, c), a), rng.plus(rng.times(b, a), rng.times(c, a)))
     | }

scala> bar(a,b,c)
res9: (scala.math.BigDecimal, scala.math.BigDecimal, scala.math.BigDecimal, scala.math.BigDecimal) = (986583317320662231.5814891877975833,986583317320662231.581489187797583304121164227599838248,986583317320662231.5814891877975833,986583317320662231.581489187797583304121164227599838248)
```

How are those situations dealt with?

Reply:
you can set up a custom `Equiv` instance:

```scala
implicit val closeEnough: Equiv[BigDecimal] = new Equiv[BigDecimal] {
  val eps = BigDecimal(1e-10) // or something.
  def equiv(a: BigDecimal, b: BigDecimal) = 2 * (b - a).abs / (b.abs + a.abs) < eps
}
```